File: _143_reorder_list/main.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {
        // Find the middle of the list
        ListNode* slow = head;
        ListNode* fast = head->next;

        while (fast != NULL && fast->next != NULL)
        {
            slow = slow->next;
            fast = fast->next->next;
        }

        // Reverse second part of the list
        ListNode* prev = NULL;
        ListNode* next = NULL;
        ListNode* current = slow->next;

        while (current != NULL)
        {
            next = current->next;
            current->next = prev;
            prev = current;
            current = next;
        }

        // Unlink two lists
        slow->next = NULL;

        // Merge two lists
        ListNode* firstList = head;
        ListNode* secondList = prev;
        ListNode* firstListNext = NULL;
        ListNode* secondListNext = NULL;

        while (secondList != NULL)
        {
            firstListNext = firstList->next;
            secondListNext = secondList->next;
            firstList->next = secondList;
            secondList->next = firstListNext;
            firstList = firstListNext;
            secondList = secondListNext;
        }
    }
};

File: _143_reorder_list/main.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        # Find the middle
        fast = head.next
        slow = head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        # Reverse second part of the list
        current = slow.next
        prev = None

        while current:
            tmp = current.next
            current.next = prev
            prev = current
            current = tmp

        # Unlink two lists
        slow.next = None

        # Merge two lists
        firstList = head
        secondList = prev

        while secondList != None:
            tmp1 = firstList.next
            tmp2 = secondList.next
            firstList.next = secondList
            secondList.next = tmp1
            firstList = tmp1
            secondList = tmp2